lua/nvim-treesitter/fold.lua

```diff
@@ -40,7 +40,12 @@ local folds_levels = tsutils.memoize_by_buf_tick(function(bufnr)
   local min_fold_lines = api.nvim_win_get_option(0, "foldminlines")
 
   for _, node in ipairs(matches) do
-    local start, _, stop, stop_col = node.node:range()
+    local start, stop, stop_col
+    if node.metadata and node.metadata.range then
```

Comment:
a bad thing about this `metadata.XXX` is that it is not specific to a capture (there could be multiple within one query). I had this doubt when we first added `metadata.priority` but then concluded that for priority we could also apply it to all the captures in query. For `range` I would  expect that it its unambiguous to what it refers to especially since `trim!` has the capture as an argument.

Apart from that this is a very important feature that could be used to replace `lua/nvim-treesitter/tsrange.lua` and by that also solve the performance problems of nvim-treesitter. I think `range` feature has potential to generalize (maybe even to core after it is battle tested in the queries here)

Reply:
Did clason's comment much later. I think this approach could be used also for `offset!` to be able to use offset also for other query types.

Reply:
Again as with offset: it would probably be more convenient if such queries would replace/modify the found nodes instead of setting metadata with the goal that there is only one source of truth for the node range

Reply:
> ... and by that also solve the performance problems of nvim-treesitter.

Interesting. Why do you think this solves the perf issues?

> I think this approach could be used also for offset! to be able to use offset also for other query types.

I based this on `offset!` which simply populates `metadata.content`. I'm not sure what I've done different here.

> Again as with offset: it would probably be more convenient if such queries would replace/modify the found nodes instead of setting metadata with the goal that there is only one source of truth for the node range

Agreed, I simply took the route of the least resistance here.

So what's needed here? Is the change good as is (...for now), or do we need to approach this differently. I only know half of what I'm doing, so I don't doubt there might be some better or simpler ways of doing this.

Reply:
We definitely need a solution that works for multiple captures in a query. Otherwise, `range` in meta data would probably work perfectly and could work the same way also for offset! (to use offset! outside of the injection world)

Regarding perf, there is nothing bad with lua/nvim-treesitter/tsrange.lua. the problem is that `make-range!` (the only directive that uses it) doesn't use the new metadata API but the an old abstraction of nvim-treesitter's query API which is aloe because it constructs a high number of tables.